<a href="https://colab.research.google.com/github/kashaf11303/urdu-ocr-codesaviours-si26-kashaf/blob/main/SI26_Week4_kashaf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install transformers datasets sentencepiece accelerate -q

In [3]:
import os
import pandas as pd
import torch

from PIL import Image
from torch.utils.data import Dataset, DataLoader

from transformers import (
    TrOCRProcessor,
    VisionEncoderDecoderModel
)

In [4]:
csv_path = "/content/drive/MyDrive/labels_200.csv"

df = pd.read_csv(csv_path)

print(df.head())
print("Rows:", len(df))

                                      image                           text
0   /content/drive/MyDrive/200 images/1.png      ہوٹل کا کھانا مزے کا تھا۔
1   /content/drive/MyDrive/200 images/2.png             یہاں کمرے میں چائے
2  /content/drive/MyDrive/200 images/20.png                  نے پلائی گئی۔
3   /content/drive/MyDrive/200 images/3.png         سٹاف نہایت دوستانہ ہے۔
4   /content/drive/MyDrive/200 images/4.png  ہوٹل کا واش روم صاف نہیں تھا۔
Rows: 200


In [6]:
missing = []

for p in df["image"]:
    if not os.path.exists(p):
        missing.append(p)

print("Missing:", len(missing))

Missing: 0


In [7]:
!pip install -U transformers sentencepiece tokenizers tiktoken
!pip uninstall -y transformers tokenizers sentencepiece tiktoken
!pip install transformers==4.44.2 tokenizers==0.19.1 sentencepiece tiktoken

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 98.0 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 5.13.1
    Uninstalling transformers-5.13.1:
      Successfully uninstalled transformers-5.13.1


Found existing installation: transformers 5.14.1
Uninstalling transformers-5.14.1:
  Successfully uninstalled transformers-5.14.1
Found existing installation: tokenizers 0.22.2
Uninstalling tokenizers-0.22.2:
  Successfully uninstalled tokenizers-0.22.2
Found existing installation: sentencepiece 0.2.2
Uninstalling sentencepiece-0.2.2:
  Successfully uninstalled sentencepiece-0.2.2
Found existing installation: tiktoken 0.13.0
Uninstalling tiktoken-0.13.0:
  Successfully uninstalled tiktoken-0.13.0
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 115.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 119.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 76.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 65.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 34.2 MB/s eta 0:00:00
  Attempting uninstall: hug

KeyboardInterrupt: 

In [1]:
import transformers
import sentencepiece
import tokenizers

print("Transformers:", transformers.__version__)
print("SentencePiece OK")
print("Tokenizers OK")

Transformers: 4.44.2
SentencePiece OK
Tokenizers OK


In [5]:
from transformers import TrOCRProcessor

processor = TrOCRProcessor.from_pretrained(
    "microsoft/trocr-base-handwritten"
)

print("Processor Loaded Successfully")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


preprocessor_config.json:   0%|          | 0.00/224 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/772 [00:00<?, ?B/s]

/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


Processor Loaded Successfully


In [6]:
import torch
from transformers import VisionEncoderDecoderModel

# Define device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Load model
model = VisionEncoderDecoderModel.from_pretrained(
    "microsoft/trocr-base-handwritten"
)

# Move model to device
model.to(device)

print("Model Loaded on:", device)

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of VisionEncoderDecoderModel were not initialized from the model checkpoint at microsoft/trocr-base-handwritten and are newly initialized: ['encoder.pooler.dense.bias', 'encoder.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


generation_config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Model Loaded on: cuda


In [7]:
import torch
from torch.utils.data import Dataset, DataLoader

In [8]:
class UrduOCRDataset(Dataset):

    def __init__(self, dataframe, processor):
        self.dataframe = dataframe
        self.processor = processor

    def __len__(self):
        return len(self.dataframe)

    def __getitem__(self, idx):

        image_path = self.dataframe.iloc[idx]["image"]
        text = str(self.dataframe.iloc[idx]["text"])

        image = Image.open(image_path).convert("RGB")

        pixel_values = self.processor(
            images=image,
            return_tensors="pt"
        ).pixel_values.squeeze()

        labels = self.processor.tokenizer(
            text,
            padding="max_length",
            max_length=64,
            truncation=True,
            return_tensors="pt"
        ).input_ids.squeeze()

        return {
            "pixel_values": pixel_values,
            "labels": labels
        }

In [9]:
import pandas as pd

df = pd.read_csv(csv_path)

In [10]:
from sklearn.model_selection import train_test_split

train_df, val_df = train_test_split(
    df,
    test_size=0.2,
    random_state=42
)

In [11]:
train_dataset = UrduOCRDataset(train_df, processor)

train_loader = DataLoader(
    train_dataset,
    batch_size=4,
    shuffle=True
)

In [12]:
val_dataset = UrduOCRDataset(val_df, processor)

test_loader = DataLoader(
    val_dataset,
    batch_size=4,
    shuffle=False
)

print("Test Loader Created")

Test Loader Created


In [ ]:
print(len(test_loader))

10


In [13]:
model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
model.config.pad_token_id = processor.tokenizer.pad_token_id
model.config.eos_token_id = processor.tokenizer.sep_token_id

In [14]:
model.config.decoder_start_token_id = processor.tokenizer.bos_token_id

In [15]:
import torch

optimizer = torch.optim.AdamW(model.parameters(), lr=5e-5)

In [16]:
num_epochs = 3

for epoch in range(num_epochs):

    model.train()

    total_loss = 0

    print(f"\nEpoch {epoch+1}/{num_epochs}")
    print("-"*30)

    for batch_idx, batch in enumerate(train_loader):

        pixel_values = batch["pixel_values"].to(device)
        labels = batch["labels"].to(device)

        outputs = model(
            pixel_values=pixel_values,
            labels=labels
        )

        loss = outputs.loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

        if batch_idx % 10 == 0:
            print(
                f"Batch {batch_idx}/{len(train_loader)} | Loss: {loss.item():.4f}"
            )

    avg_loss = total_loss / len(train_loader)

    print(
        f"Epoch {epoch+1} complete | Average Loss: {avg_loss:.4f}"
    )

print("Training Complete")


Epoch 1/3
------------------------------
Batch 0/40 | Loss: 16.8823
Batch 10/40 | Loss: 4.0813
Batch 20/40 | Loss: 3.5045
Batch 30/40 | Loss: 3.7415
Epoch 1 complete | Average Loss: 4.6778

Epoch 2/3
------------------------------
Batch 0/40 | Loss: 3.7109
Batch 10/40 | Loss: 3.4441
Batch 20/40 | Loss: 3.0838
Batch 30/40 | Loss: 2.5385
Epoch 2 complete | Average Loss: 3.3435

Epoch 3/3
------------------------------
Batch 0/40 | Loss: 3.0087
Batch 10/40 | Loss: 3.1964
Batch 20/40 | Loss: 2.8296
Batch 30/40 | Loss: 3.7821
Epoch 3 complete | Average Loss: 3.1404
Training Complete


In [17]:
model.eval()

print("=== Model Evaluation on Test Images ===")
print()

correct = 0
total = 0

with torch.no_grad():
    for batch in test_loader:

        pixel_values = batch["pixel_values"].to(device)
        labels = batch["labels"]

        # Generate predictions
        generated_ids = model.generate(pixel_values)

        generated_text = processor.batch_decode(
            generated_ids,
            skip_special_tokens=True
        )

        actual_text = processor.batch_decode(
            labels,
            skip_special_tokens=True
        )

        # Debug Output
        print("Generated IDs:", generated_ids)
        print("Generated Text:", generated_text)
        print("Actual Text:", actual_text)
        print("-" * 50)

        for pred, actual in zip(generated_text, actual_text):
            total += 1

            if pred.strip() == actual.strip():
                correct += 1

            print(f"Predicted: {pred}")
            print(f"Actual:    {actual}")
            print()

accuracy = (correct / total) * 100 if total > 0 else 0

print(f"Accuracy: {accuracy:.1f}% ({correct}/{total} correct)")

=== Model Evaluation on Test Images ===



/usr/local/lib/python3.12/dist-packages/transformers/generation/utils.py:1258: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


Generated IDs: tensor([[2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]],
       device='cuda:0')
Generated Text: ['', '', '', '']
Actual Text: ['شیخ چلی نے انگریزی سیکھی', 'کا', 'کروایا۔ میں دوبارہ اس ہوٹل میں نہیں آؤں گا۔', 'خاموش رہو']
--------------------------------------------------
Predicted: 
Actual:    شیخ چلی نے انگریزی سیکھی

Predicted: 
Actual:    کا

Predicted: 
Actual:    کروایا۔ میں دوبارہ اس ہوٹل میں نہیں آؤں گا۔

Predicted: 
Actual:    خاموش رہو

Generated IDs: tensor([[2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]],
       device=

In [18]:
from google.colab import drive
drive.mount('/content/drive')

save_path = '/content/drive/MyDrive/SI26-urdu-ocr-model'

model.save_pretrained(save_path)
processor.save_pretrained(save_path)

print(f'Model saved to Google Drive: {save_path}')
print('You can load this model again next week without retraining')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Model saved to Google Drive: /content/drive/MyDrive/SI26-urdu-ocr-model
You can load this model again next week without retraining


my accuracy is 0.0% and training loss went from
Epoch 1 complete | Average Loss: 4.6778
Epoch 2 complete | Average Loss: 3.3435
Epoch 3 complete | Average Loss: 3.1404